model/riscv_vmem.sail

```diff
@@ -0,0 +1,854 @@
+               ext_ptw)                      // ext_ptw
+              -> PTW_Result
+
+function pt_walk(sv_params,
+                 va,
+                 ac,
+                 priv,
+                 mxr,
+                 do_sum,
+                 pt_base,
+                 level,
+                 global,
+                 ext_ptw) = {
+  let vpn_j      = vpn_j_of_va(sv_params, va, level);
+  let pte_offset = vpn_j << sv_params.log_pte_size_bytes;
+  let pte_addr   = pt_base + pte_offset;
+
+  // In Sv32, physical addrs are actually 34 bits, not XLEN(=32) bits.
+  // Below, 'pte_phys_addr' is XLEN bits because it's an arg to
+  // 'mem_read_priv()' [riscv_mem.sail] where it's declared as xlenbits.
+  // That def and this use need to be fixed together (TODO)
+  let pte_phys_addr : xlenbits = pte_addr[(sizeof(xlen) - 1) .. 0];
+
+  // Read this-level PTE from mem
+  let mem_result = mem_read_priv(Read(Data),              // AccessType
+                                 Supervisor,              // Privilege
+                                 pte_phys_addr,
+                                 8,                       // atom (8)
+                                 false,                   // aq
+                                 false,                   // rl
+                                 false);                  // res
+
+  match mem_result {
+    MemException(_) => PTW_Failure(PTW_Access(), ext_ptw),
+    MemValue(pte)   => {
+      let pte_lsbs = Mk_PTE_LSBs(pte[7 .. 0]);
+      if pte_is_invalid(pte_lsbs) then
+        PTW_Failure(PTW_Invalid_PTE(), ext_ptw)
+      else {
+        // Non-Leaf PTE
+        let ppns : bits(64) = PPNs_of_PTE(sv_params, pte);
+        let global'         = global | (pte_lsbs.G() == 0b1);
+        if pte_is_ptr(pte_lsbs) then {
+          if level > 0 then {
+            // follow the pointer to walk next level
+            let pt_base' : bits(64) = ppns << pagesize_bits;
+            let level'              = level - 1;
+            pt_walk(sv_params, va, ac, priv, mxr, do_sum,
+                    pt_base', level', global', ext_ptw)
+          }
+          else
+            // level 0 PTE, but contains a pointer instead of a leaf
+            PTW_Failure(PTW_Invalid_PTE(), ext_ptw)
+        }
+        // Leaf PTE
+        else {
+          let ext_pte   = msbs_of_PTE(sv_params, pte);
+          let pte_check = check_PTE_permission(ac, priv, mxr, do_sum, pte_lsbs,
+                                               ext_pte, ext_ptw);
+          match pte_check {
+            PTE_Check_Failure(ext_ptw, ext_ptw_fail) =>
+              PTW_Failure(ext_get_ptw_error(ext_ptw_fail), ext_ptw),
+            PTE_Check_Success(ext_ptw) =>
+              if level > 0 then {
+                // Superpage; construct mask for lower-level PPNs from the PTE
+                let mask_PPN_j : bits(64) = zero_extend(ones(sv_params.pte_PPN_j_size_bits));
+                mask           : bits(64) = zeros();
```

Comment:
`var mask` (implicit `var` will be an error eventually).

Reply:
Ok. I was unaware that Sail now recommends (and soon may demand) ‘var’ for updatable variables. But this fix is moot because of the reformulation in the next suggestion, which we’ve adopted, and which has no vars.